### PR TITLE
Improve CSPLayout performance by reducing constraint complexity

### PR DESCRIPTION
**Pull request by 1ucian0**
Based on ideas after talking with @czachow, it was possible to *specialised* the `CSPLayout` pass and to reduce the complexity of the constraints to check. This gives an improvement in the performance. How much? It varies a lot, depending on the specific situation. Here is an example (based on https://github.com/Qiskit/qiskit-terra/issues/5694) where the goal is to find the perfect solution in `Toronto` (100 times, to amplify the result):

```python
from qiskit.circuit.library import GraphState
from qiskit.test.mock import FakeToronto
import numpy as np
from qiskit.transpiler.passes import CSPLayout
from qiskit.transpiler import CouplingMap
from qiskit.converters import circuit_to_dag
from time import time

backend = FakeToronto()
config = backend.configuration()

rows = [x[0] for x in config.coupling_map]
cols = [x[1] for x in config.coupling_map]

A = np.zeros((config.num_qubits, config.num_qubits))
A[rows, cols] = 1

qc = GraphState(A)
qc.measure_all()

dag = circuit_to_dag(qc)
start = time()
for i in range(100):
    pass_ = CSPLayout(CouplingMap(config.coupling_map), seed=i, call_limit=None, time_limit=None)
    pass_.run(dag)
print(time()-start)
print(pass_.property_set['CSPLayout_stop_reason'])
```

The current result is: 342.403 sec

With this PR: ~40.175~  241.838 sec (see https://github.com/Qiskit/qiskit-terra/pull/6263#issuecomment-829333293)

### CODE REVIEW

**Comment by 1ucian0** [comment]
@czachow found a case we the solution was not being found (test added in 3b7c991ce97ccd22c155735367ed5397ce0a563a). As a consequence more possible assignments need to be checked that increases run time.

File: qiskit/transpiler/passes/layout/csp_layout.py [hunk 1]
```diff
@@ -109,6 +136,10 @@ def run(self, dag):
                      qubits.index(gate.qargs[1])))
         edges = set(self.coupling_map.get_edges())
 
+        if not self.strict_direction:
+            cxs = {frozenset(cx) for cx in cxs}
```
Comment (mtreinish) [inline qiskit/transpiler/passes/layout/csp_layout.py:140]: This feel unnecessary as you end up iterating over the set of two qubit ops twice. Why not just have the if block around the loop and do this iteratively as you build cx?

File: qiskit/transpiler/passes/layout/csp_layout.py [hunk 2]
```diff
@@ -109,6 +136,10 @@ def run(self, dag):
...
+            edges = {frozenset(edge) for edge in edges}
```
Comment (mtreinish) [inline qiskit/transpiler/passes/layout/csp_layout.py:141]: Can you open a retworkx issue here on this https://github.com/Qiskit/retworkx/issues/new/choose I feel like there should be a native method or function on the `PyDiGraph` class that` CouplingMap` wraps to return the set of weakly connected edges for a graph. It'd be a bit faster too if that method exists (you might be able to hack it together inside couplingmap today with `.to_undirected` https://qiskit.org/documentation/retworkx/stubs/retworkx.PyDiGraph.to_undirected.html#retworkx.PyDiGraph.to_undirected but having a native method would be best.

File: qiskit/transpiler/passes/layout/csp_layout.py [hunk 3]
```diff
@@ -65,6 +66,32 @@ def recursiveBacktracking(self,  # pylint: disable=invalid-name
                                              single)
 
 
+class CustomConstraint(Constraint):
```
Comment (mtreinish) [inline qiskit/transpiler/passes/layout/csp_layout.py:69]: Maybe rename this to be a bit more descriptive, something like `ValidEdgeConstrait` or something like that.

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, just a couple inline suggestions nothing big though. Also I think there is a missing feature from retworkx that this exposes if you could open an issue about it there too that would be great.

**Comment by 1ucian0** [comment]
With VF2 taking traction, I think this PR can be closed.

**Pull request closed without merging**